peekingduck/pipeline/nodes/heuristic/fps.py

```diff
@@ -0,0 +1,86 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Calculates the FPS of video
+"""
+
+from typing import Any, Dict, List
+from time import perf_counter
+import numpy as np
+
+from peekingduck.pipeline.nodes.node import AbstractNode
+
+NUM_FRAMES = 14
+
+
+class Node(AbstractNode):
+    """ FPS node class that calculates the FPS of the image frame """
+
+    def __init__(self, config: Dict[str, Any]) -> None:
+        super().__init__(config, node_path=__name__)
+        self.time_window = [float(0)]
+
+        self.moving_avg_thres = config["moving_avg"]
+        self.moving_average_fps: List[float] = []
+        self.global_avg_fps = 0.0
+        self.count = 0
+
+    def run(self, inputs: Dict[str, Any]) -> Dict[str, Any]:
+        """ Calculates FPS using the time difference between the current
+        frame and the previous frame.
+
+        Args:
+            inputs: ["pipeline_end"]
+
+        Returns:
+            outputs (dict): Dict with key "fps".
+        """
+
+        if len(self.time_window) > NUM_FRAMES:
+            self.time_window.pop(0)
+
+        self.time_window.append(perf_counter())
+
+        num_frames = len(self.time_window)
+        time_diff = self.time_window[-1] - self.time_window[0]
+        average_fps = num_frames / time_diff
+
+        # Logs 14-frame moving average per frame
+        if self.moving_avg_thres:
+            moving_average = self._moving_average(average_fps)
+            self.logger.info('14-frame Moving Average FPS: %s', moving_average)
+
+        # Calculate global cumulative moving average
+        self.global_avg_fps = self._global_average(average_fps)
+        # Log global cumulative average when pipeline ends
+        if inputs["pipeline_end"]:
+            self.logger.info('Approximate Global Average FPS: %s',
+                self.global_avg_fps)
+
+        return {"fps": average_fps}
+
+    def _moving_average(self, average_fps: float) -> float:
+        self.moving_average_fps.append(average_fps)
+        if len(self.moving_average_fps) > NUM_FRAMES:
+            self.moving_average_fps.pop(0)
+        moving_average_val = np.mean(self.moving_average_fps)
+        return round(moving_average_val, 1)
+
+    def _global_average(self, average_fps: float) -> float:
+        # Cumulative moving average formula
+        global_average = \
+            (average_fps + self.count*self.global_avg_fps) / (self.count + 1)
+        self.count += 1
+        return round(global_average, 1)
```

Comment:
What's causing the discrepancy?

Let's take a look at the printout of `self.time_window`:

```
2021-07-16T17:10:21 peekingduck.loaders INFO:Initialising output.screen node...
[0.0, 6.985116089]
2021-07-16T17:10:21 peekingduck.pipeline.nodes.heuristic.fps INFO:14-frame Moving Average FPS: 0.3
[0.0, 6.985116089, 8.039921208]
2021-07-16T17:10:23 peekingduck.pipeline.nodes.heuristic.fps INFO:14-frame Moving Average FPS: 0.3
[0.0, 6.985116089, 8.039921208, 8.101609373]
2021-07-16T17:10:23 peekingduck.pipeline.nodes.heuristic.fps INFO:14-frame Moving Average FPS: 0.4
[0.0, 6.985116089, 8.039921208, 8.101609373, 8.16379758]
2021-07-16T17:10:23 peekingduck.pipeline.nodes.heuristic.fps INFO:14-frame Moving Average FPS: 0.4
[0.0, 6.985116089, 8.039921208, 8.101609373, 8.16379758, 8.217789053]
2021-07-16T17:10:23 peekingduck.pipeline.nodes.heuristic.fps INFO:14-frame Moving Average FPS: 0.5
[0.0, 6.985116089, 8.039921208, 8.101609373, 8.16379758, 8.217789053, 8.279178602]
2021-07-16T17:10:23 peekingduck.pipeline.nodes.heuristic.fps INFO:14-frame Moving Average FPS: 0.6
[0.0, 6.985116089, 8.039921208, 8.101609373, 8.16379758, 8.217789053, 8.279178602, 8.345702212]
2021-07-16T17:10:23 peekingduck.pipeline.nodes.heuristic.fps INFO:14-frame Moving Average FPS: 0.6
[0.0, 6.985116089, 8.039921208, 8.101609373, 8.16379758, 8.217789053, 8.279178602, 8.345702212, 8.403459725]
2021-07-16T17:10:23 peekingduck.pipeline.nodes.heuristic.fps INFO:14-frame Moving Average FPS: 0.7
[0.0, 6.985116089, 8.039921208, 8.101609373, 8.16379758, 8.217789053, 8.279178602, 8.345702212, 8.403459725, 8.466674783]
2021-07-16T17:10:23 peekingduck.pipeline.nodes.heuristic.fps INFO:14-frame Moving Average FPS: 0.7
[0.0, 6.985116089, 8.039921208, 8.101609373, 8.16379758, 8.217789053, 8.279178602, 8.345702212, 8.403459725, 8.466674783, 8.530684311]
```

The formula: time_diff = self.time_window[-1] - self.time_window[0]
Causes the zero at the beginning of the list to be used for the calculation (self.time_window[0]) all the time until it gets popped out by the 14th frame. The value of zero is vastly different from the rest of the values in self.time_window, which is why it has a pronounced effect for a short run.

Reply:
2 things you could try now:
1) Try to negate or minimise the effect of the first zero by tweaking the formula
2) If (1) is not possible, will have to use the instantaneous fps from my code snippet above. In this case, if `draw.fps` were to read "fps" from the instantaneous fps, it'll be very jittery, ranging from 0.9 to 15 to 7 to 36. Perhaps we'll need to revert draw.fps to the previous version which calculates a "dampened" fps that reduces this effect.

Reply:
I have added the instantaneous fps calculation into the heuristic. This instantaneous FPS value will be used for the moving average and global cumulative average calculations. I have kept the average_fps from earlier because this can be the dampened FPS and this value is the Node's output.